File: ios-app/GalaxyBridge/Bridge/BridgeProtocol.swift
import Foundation

struct BridgeEnvelope: Identifiable, Equatable {
    let id: String
    let type: String
    let timestampSeconds: Int
    let payload: [String: Any]
    let version: Int

    init(
        id: String = UUID().uuidString,
        type: String,
        timestampSeconds: Int = Int(Date().timeIntervalSince1970),
        payload: [String: Any],
        version: Int = BridgeGattProfile.protocolVersion
    ) {
        self.id = id
        self.type = type
        self.timestampSeconds = timestampSeconds
        self.payload = payload
        self.version = version
    }

    static func == (lhs: BridgeEnvelope, rhs: BridgeEnvelope) -> Bool {
        lhs.id == rhs.id && lhs.type == rhs.type
    }
}

enum BridgeProtocol {
    static let maxChunkBytes = 160

    static func encode(_ envelope: BridgeEnvelope) throws -> Data {
        let object: [String: Any] = [
            "v": envelope.version,
            "id": envelope.id,
            "type": envelope.type,
            "ts": envelope.timestampSeconds,
            "payload": envelope.payload
        ]
        return try JSONSerialization.data(withJSONObject: object)
    }

    static func decode(_ data: Data) throws -> BridgeEnvelope {
        let object = try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]
        return BridgeEnvelope(
            id: object["id"] as? String ?? UUID().uuidString,
            type: object["type"] as? String ?? "unknown",
            timestampSeconds: object["ts"] as? Int ?? Int(Date().timeIntervalSince1970),
            payload: object["payload"] as? [String: Any] ?? [:],
            version: object["v"] as? Int ?? 1
        )
    }

    static func ack(for id: String) -> BridgeEnvelope {
        BridgeEnvelope(type: "ack", payload: ["for": id])
    }

    static func chunks(for envelope: BridgeEnvelope) throws -> [Data] {
        let data = try encode(envelope)
        guard data.count > maxChunkBytes else { return [data] }
        let total = Int(ceil(Double(data.count) / Double(maxChunkBytes)))
        return stride(from: 0, to: data.count, by: maxChunkBytes).enumerated().map { index, offset in
            let end = min(offset + maxChunkBytes, data.count)
            let part = data.subdata(in: offset..<end)
            let object: [String: Any] = [
                "chunk": true,
                "id": envelope.id,
                "seq": index,
                "total": total,
                "data": part.base64EncodedString()
            ]
            return try! JSONSerialization.data(withJSONObject: object)
        }
    }
}

final class ChunkReassembler {
    private var chunks: [String: [Int: String]] = [:]
    private var totals: [String: Int] = [:]
    private var seen: [String] = []
    private var seenSet: Set<String> = []

    func remember(_ id: String) -> Bool {
        guard !seenSet.contains(id) else { return false }
        seen.append(id)
        seenSet.insert(id)
        while seen.count > 64 {
            seenSet.remove(seen.removeFirst())
        }
        return true
    }

    func accept(_ data: Data) throws -> Data? {
        let object = try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]
        guard object["chunk"] as? Bool == true else { return data }
        guard
            let id = object["id"] as? String,
            let seq = object["seq"] as? Int,
            let total = object["total"] as? Int,
            let fragment = object["data"] as? String
        else { return nil }

        totals[id] = total
        var map = chunks[id, default: [:]]
        map[seq] = fragment
        chunks[id] = map
        guard map.count == total else { return nil }

        var merged = Data()
        for index in 0..<total {
            guard let fragment = map[index], let data = Data(base64Encoded: fragment) else {
                return nil
            }
            merged.append(data)
        }
        chunks[id] = nil
        totals[id] = nil
        return merged
    }
}

File: watch-app/app/src/main/java/com/localbridge/watch/ble/BridgeProtocol.kt
package com.localbridge.watch.ble

import android.util.Base64
import android.util.Log
import org.json.JSONObject
import java.nio.charset.StandardCharsets
import java.util.UUID
import kotlin.math.ceil

data class BridgeEnvelope(
    val id: String = UUID.randomUUID().toString(),
    val type: String,
    val timestampSeconds: Long = System.currentTimeMillis() / 1000,
    val payload: JSONObject = JSONObject(),
    val version: Int = BridgeGattProfile.PROTOCOL_VERSION
) {
    fun toJson(): JSONObject = JSONObject()
        .put("v", version)
        .put("id", id)
        .put("type", type)
        .put("ts", timestampSeconds)
        .put("payload", payload)

    fun toBytes(): ByteArray = toJson().toString().toByteArray(StandardCharsets.UTF_8)
}

object BridgeProtocol {
    const val MAX_CHUNK_BYTES = 160

    fun decode(bytes: ByteArray): BridgeEnvelope {
        val json = JSONObject(bytes.toString(StandardCharsets.UTF_8))
        return BridgeEnvelope(
            id = json.getString("id"),
            type = json.getString("type"),
            timestampSeconds = json.optLong("ts", System.currentTimeMillis() / 1000),
            payload = json.optJSONObject("payload") ?: JSONObject(),
            version = json.optInt("v", 1)
        )
    }

    fun ack(forId: String) = BridgeEnvelope(
        type = "ack",
        payload = JSONObject().put("for", forId)
    )

    fun chunk(envelope: BridgeEnvelope): List<ByteArray> {
        val text = envelope.toJson().toString()
        val bytes = text.toByteArray(StandardCharsets.UTF_8)
        if (bytes.size <= MAX_CHUNK_BYTES) return listOf(bytes)

        val total = ceil(bytes.size / MAX_CHUNK_BYTES.toDouble()).toInt()
        return bytes.asList().chunked(MAX_CHUNK_BYTES).mapIndexed { index, part ->
            JSONObject()
                .put("chunk", true)
                .put("id", envelope.id)
                .put("seq", index)
                .put("total", total)
                .put("data", Base64.encodeToString(part.toByteArray(), Base64.NO_WRAP))
                .toString()
                .toByteArray(StandardCharsets.UTF_8)
        }
    }
}

class ChunkReassembler {
    private val chunks = mutableMapOf<String, MutableMap<Int, String>>()
    private val totals = mutableMapOf<String, Int>()
    private val duplicateSeqCounts = mutableMapOf<String, Int>()
    private val droppedIds = mutableSetOf<String>()
    private val seen = ArrayDeque<String>()
    private val seenSet = mutableSetOf<String>()

    fun remember(id: String): Boolean {
        if (seenSet.contains(id)) return false
        seen.addLast(id)
        seenSet.add(id)
        while (seen.size > 64) seenSet.remove(seen.removeFirst())
        return true
    }

    fun accept(bytes: ByteArray): ByteArray? {
        val json = JSONObject(bytes.toString(StandardCharsets.UTF_8))
        if (!json.optBoolean("chunk", false)) return bytes

        val id = json.getString("id")
        if (droppedIds.contains(id)) {
            Log.w(TAG, "WRITE_CHUNK_DROPPED id=$id reason=previous_duplicate_drop")
            return null
        }
        val total = json.getInt("total")
        val seq = json.getInt("seq")
        Log.i(TAG, "WRITE_CHUNK id=$id seq=$seq total=$total")
        totals[id] = total
        val map = chunks.getOrPut(id) { mutableMapOf() }
        if (map.containsKey(seq)) {
            val count = (duplicateSeqCounts[id] ?: 0) + 1
            duplicateSeqCounts[id] = count
            Log.w(TAG, "WRITE_CHUNK_DUPLICATE id=$id seq=$seq count=$count")
            if (count > 3) {
                chunks.remove(id)
                totals.remove(id)
                duplicateSeqCounts.remove(id)
                droppedIds.add(id)
                Log.w(TAG, "WRITE_CHUNK_DUPLICATE_DROP id=$id seq=$seq total=$total")
            }
            return null
        }
        map[seq] = json.getString("data")
        if (map.size != total) return null

        val merged = (0 until total).fold(ByteArray(0)) { acc, index ->
            acc + Base64.decode(map.getValue(index), Base64.NO_WRAP)
        }
        chunks.remove(id)
        totals.remove(id)
        duplicateSeqCounts.remove(id)
        droppedIds.remove(id)
        return merged
    }

    companion object {
        private const val TAG = "GalaxyBridgeBle"
    }
}
